src/raftstore/store/worker/region.rs

```diff
@@ -305,24 +441,169 @@ impl Runnable<Task> for Runner {
                 region_id,
                 notifier,
             } => {
-                // It safe for now to handle generating and applying snapshot concurrently,
+                // It is safe for now to handle generating and applying snapshot concurrently,
                 // but it may not when merge is implemented.
                 let ctx = self.ctx.clone();
                 self.pool
                     .execute(move |_| ctx.handle_gen(region_id, notifier))
             }
-            Task::Apply { region_id, status } => self.ctx.handle_apply(region_id, status),
+            Task::Apply { region_id, status } => {
+                self.ctx
+                    .handle_apply(&mut self.regions_to_delete, region_id, status)
+            }
             Task::Destroy {
                 region_id,
                 start_key,
                 end_key,
-            } => self.ctx.handle_destroy(region_id, start_key, end_key),
+            } => {
+                info!(
+                    "[region {}] register deleting data in [{}, {})",
+                    region_id,
+                    escape(&start_key),
+                    escape(&end_key)
+                );
+                self.regions_to_delete.update(start_key, end_key, true);
```

Comment:
Don't use `regions_to_delete` directly here, abstract a `delay_destroy` function.

Reply:
And give an explanation why delay is needed.